### PR TITLE
Implement MultioutputWrapper

### PR DESCRIPTION
**Pull request by an1lam**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?

Implements the `MultioutputWrapper` feature described in #508. Summarizing, `MultioutputWrapper` takes a base metric and a `num_outputs` param and returns a metric that will compute `shape[output_dim]` many scalars and return them in a list. The idea is to support `multioutput` mode for metrics like `SpearmanCorrcoef` that don't support it natively.

### CODE REVIEW

**Comment by an1lam** [comment]
I'm not sure why the build is failing but it seems to be totally independent of my change & instead has to do with Docker. It looks like it's also failing on #506, so I'm assuming this is not caused by my change?

File: torchmetrics/wrappers/multioutput.py
Comment (SkafteNicki) [inline torchmetrics/wrappers/multioutput.py:62]: some examples in the end would be nice, similar to:
https://github.com/PyTorchLightning/metrics/blob/3a4ff87bdf1f0fdf394b414041e5858adc819b12/torchmetrics/regression/r2.py#L75-L87

File: torchmetrics/wrappers/multioutput.py
Comment (SkafteNicki) [inline torchmetrics/wrappers/multioutput.py:92]: I am not sure if this will give device errors since index should can be on different device than the input
```suggestion
                args, torch.Tensor, torch.index_select, dim=self.output_dim, index=torch.tensor(i, device=self.device)
            )
            selected_kwargs = apply_to_collection(
                kwargs, torch.Tensor, torch.index_select, dim=self.output_dim, index=torch.tensor(i, device=self.device)
```

File: tests/wrappers/test_multioutput.py
Comment (SkafteNicki) [inline tests/wrappers/test_multioutput.py:8]: For testing it would be great if we tested on random input instead of specific

File: torchmetrics/wrappers/multioutput.py
Comment (SkafteNicki) [inline torchmetrics/wrappers/multioutput.py:67]: I would make this a required parameter since it is so essential for what the class is doing
```suggestion
        num_outputs: int,
```

**Review by SkafteNicki — COMMENTED** [review]
Few comments, PR is looking good already :]

File: torchmetrics/wrappers/multioutput.py
Comment (an1lam) [reply torchmetrics/wrappers/multioutput.py]: Done!

File: torchmetrics/wrappers/multioutput.py
Comment (an1lam) [reply torchmetrics/wrappers/multioutput.py]: Done!

File: torchmetrics/wrappers/multioutput.py
Comment (an1lam) [reply torchmetrics/wrappers/multioutput.py:89]: Added two, let me know if these seem good/are enough.

File: tests/wrappers/test_multioutput.py
Comment (an1lam) [reply tests/wrappers/test_multioutput.py]: Done!

**Comment by an1lam** [comment]
@SkafteNicki not exactly sure at what point I move this out of draft mode/whether I am the one supposed to do it. Let me know!

**Comment by SkafteNicki** [comment]
@an1lam Changed it to review status, to get all test running.

File: tests/wrappers/test_multioutput.py
Comment (SkafteNicki) [inline tests/wrappers/test_multioutput.py:38]: Based on how things are parameterized, it looks like it can fit into our standard testing framework like this:
https://github.com/PyTorchLightning/metrics/blob/e6d998e11fc2b93286d62beaa70f6a47647d77eb/tests/regression/test_r2.py#L71-L84
It would be great if you could do this, because the wrapper should then also be tested automatically for cuda and multi device (ddp) support

**Review by SkafteNicki — COMMENTED** [review]
I am basically fine with the PR as it is, but it would be great to further improve the testing (I doubt it will show any problems as this is just a wrapper around other metrics) :]

File: tests/wrappers/test_multioutput.py
Comment (an1lam) [reply tests/wrappers/test_multioutput.py:38]: Awesome, thanks for the tip! Will do.

Comment (an1lam) [reply tests/wrappers/test_multioutput.py:38]: So in trying to get this to work, I realized it actually won't work because `forward` calls `reset` but in my current setup (and the bootstrap wrapper's by the way), setting `self.metrics = nn.ModuleList` doesn't register the underlying metrics as part of the state. As a result, `forward` will reset the state meaning `compute()` will return the result from the last batch rather than over all batches.

One idea I had was trying to make `metrics` part of state, but that won't work if I continue to use an `nn.ModuleList`. Alternatively, I was wondering if you think it makes sense to try and make the wrapper's state the union of the underlying states (with indexes to differentiate). Other ideas also welcome!

Comment (an1lam) [reply tests/wrappers/test_multioutput.py:38]: Ok going to override `forward` actually, this isn't too messy.

File: tests/wrappers/test_multioutput.py
Comment (an1lam) [inline tests/wrappers/test_multioutput.py:21]: This should say "Test class that allows passing base metric as a class rather than its instantiation to the wrapper."

File: tests/wrappers/test_multioutput.py
Comment (an1lam) [inline tests/wrappers/test_multioutput.py:71]: Use `NUM_CLASSES` from helper.

File: torchmetrics/wrappers/__init__.py
Comment (an1lam) [inline torchmetrics/wrappers/__init__.py:15]: This is the wrong order. Should go after `MetricTracker`.

**Comment by an1lam** [comment]
I'm not totally sure what to do about the failing tests on older versions. It looks like the issue is I'm using `movedim`, which isn't supported in older versions of Pytorch. Is there a decorator I can add that flags this new class as incompatible with these old versions? Is this even an acceptable solution?

**Comment by SkafteNicki** [comment]
Hi @an1lam, please check your examples, they do not seem to render correctly:
image

**Review by SkafteNicki — COMMENTED** [review]
LGTM :]

**Review by SkafteNicki — APPROVED** [review]

**Comment by an1lam** [comment]
@SkafteNicki do you think I should wait until you merge #542 then (and merge `master` afterwards) to merge?

**Comment by SkafteNicki** [comment]
@an1lam yes that PR should fix the last test that fails here :]

File: torchmetrics/wrappers/multioutput.py
Comment (Borda) [inline torchmetrics/wrappers/multioutput.py:65]:
```suggestion
        base_metric:
            Metric being wrapped.
        num_outputs:
            Expected dimensionality of the output dimension. This parameter is
            used to determine the number of distinct metrics we need to track.
        output_dim:
            Dimension on which output is expected. Note that while this provides some flexibility, the output dimension
            must be the same for all inputs to update. This applies even for metrics such as `Accuracy` where the labels
            can have a different number of dimensions than the predictions. This can be worked around if the output
            dimension can be set to -1 for both, even if -1 corresponds to different dimensions in different inputs.
        remove_nans:
            Whether to remove the intersection of rows containing NaNs from the values passed through to each underlying
            metric. Proper operation requires all tensors passed to update to have dimension `(N, ...)` where N
            represents the length of the batch or dataset being passed in.
        squeeze_outputs:
            If true, will squeeze the 1-item dimensions left after `index_select` is applied.
            This is sometimes unnecessary but harmless for metrics such as `R2Score` but useful
            for certain classification metrics that can't handle additional 1-item dimensions.
        compute_on_step:
            Whether to recompute the metric value on each update step.
        dist_sync_on_step:
            Required for distributed training support. See torchmetrics docs for additional details.
        process_group:
            Specify the process group on which synchronization is called. default: None (which selects the entire world)
        dist_sync_fn:
            Required for distributed training support. See torchmetrics docs for additional details.
```
typing and defaults are taken from sinature

File: torchmetrics/wrappers/multioutput.py
Comment (Borda) [inline torchmetrics/wrappers/multioutput.py:29]:
```suggestion
    Unlike specific torchmetric metrics, it doesn't support any aggregation across outputs. This means if you set `num_outputs` to 2, `compute()` will return a Tensor of dimension
```

File: torchmetrics/wrappers/multioutput.py
Comment (Borda) [inline torchmetrics/wrappers/multioutput.py:36]:
```suggestion
    In addition to enabling multioutput support for metrics that lack it, this class also supports, albeit in a crude
    fashion, dealing with missing labels (or other data). When ``remove_nans`` is passed, the class will remove the
    intersection of NaN containing "rows" upon each update for each output. For example, suppose a user uses
    `MultioutputWrapper` to wrap :class:`torchmetrics.regression.r2.R2Score` with 2 outputs, one of which occasionally
    has missing labels for classes like ``R2Score`` is that this class supports removing NaN values
    (parameter ``remove_nans``) on a per-output basis. When ``remove_nans`` is passed the wrapper will remove all rows
```

**Review by Borda — APPROVED** [review]

**Pull request merged**
